### PR TITLE
Implement SPIR-V target for GLSL functions

### LINKED ISSUE #4051: Use the GLSL.std.450 instruction set in favor of emulating instructions.
Currently some functions like atanh, packUnorm2x16, etc. emulate functionality when targeting a SPIR-V backend. This should be changed to use the [glsl.std.450 extended instruction set of SPIR-V for correctness & performance reasons](https://registry.khronos.org/SPIR-V/specs/unified1/GLSL.std.450.html).

### PR DESCRIPTION
**Pull request by jkwak-work**
Fixes #4051

This commit implements SPIR-V target for GLSL functions. It also fixes a few problesm of GLSL targetting implemention too.

### CODE REVIEW

**Review by tgrimesnv — APPROVED** [review]

**Comment by ArielG-NV** [comment]
hlsl.meta, `fmod` (6505) should be using SPIR-V [OpFRem](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpFRem). Note, OpFMod is not the same as fmod,

currently it does:
```
    __target_switch
    {
    case cpp: __intrinsic_asm "$P_fmod($0, $1)";
    case cuda: __intrinsic_asm "$P_fmod($0, $1)";
    case hlsl: __intrinsic_asm "fmod";
    default:
        return x - y * trunc(x/y);
    }
```

should be 
```
    __target_switch
    {
    case cpp: __intrinsic_asm "$P_fmod($0, $1)";
    case cuda: __intrinsic_asm "$P_fmod($0, $1)";
    case hlsl: __intrinsic_asm "fmod";
    case spirv: 
        return spirv_asm
        {
            OpFRem $$T result $0 $1
        }
    default:
        return x - y * trunc(x/y);
    }
```

**Comment by jkwak-work** [comment]
> hlsl.meta, `fmod` (6505) should be using SPIR-V [OpFRem](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpFRem). Note, `OpFMod` is not the same as `fmod()` in hlsl.
> 
> currently it does:
> 
> ```
>     __target_switch
>     {
>     case cpp: __intrinsic_asm "$P_fmod($0, $1)";
>     case cuda: __intrinsic_asm "$P_fmod($0, $1)";
>     case hlsl: __intrinsic_asm "fmod";
>     default:
>         return x - y * trunc(x/y);
>     }
> ```
> 
> should be
> 
> ```
>     __target_switch
>     {
>     case cpp: __intrinsic_asm "$P_fmod($0, $1)";
>     case cuda: __intrinsic_asm "$P_fmod($0, $1)";
>     case hlsl: __intrinsic_asm "fmod";
>     case spirv: 
>         return spirv_asm
>         {
>             OpFRem $$T result $0 $1
>         }
>     default:
>         return x - y * trunc(x/y);
>     }
> ```

Yes, it is in another PR separately.

**Comment by ArielG-NV** [comment]
Everything looks good 👍.

I think I will make a separate issue for vector ops, I noticed some which accept vector ops are using `VECTOR_MAP_UNARY()`

**Review by ArielG-NV — APPROVED** [review]

File: source/slang/hlsl.meta.slang [hunk 1]
```diff
@@ -6028,9 +6036,7 @@ T exp2(T x)
     __target_switch
     {
     case glsl:
-        if (__isHalf<T>())
-            __intrinsic_asm "exp2($0)";
-        __intrinsic_asm "exp2(float($0))";
+        __intrinsic_asm "exp2($0)";
```
Comment (csyonghe) [inline source/slang/hlsl.meta.slang:6039]: I think this was because exp2 isn't defined for double.

**Review by csyonghe — APPROVED** [review]

File: source/slang/hlsl.meta.slang (hunk 1 above)
Comment (jkwak-work) [reply source/slang/hlsl.meta.slang:6039]: Good catch.
It was giving me some trouble and I had to remove "float".
I may have to come back to this later.

File: tests/glsl-intrinsic/intrinsic-basic.slang [hunk 2]
```diff
@@ -1,29 +1,23 @@
...
+//TEST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl
+//TEST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target cuda -DTARGET_CUDA
+//TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
 
-//TEST:SIMPLE(filecheck=CHECK_GLSL): -allow-glsl -stage compute -entry computeMain -target glsl -g2
-//TEST:SIMPLE(filecheck=CHECK_SPV):  -allow-glsl -stage compute -entry computeMain -target spirv  -g2
-//TEST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl  -g2
-//TEST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target cuda -DTARGET_CUDA  -g2
-//TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp  -g2
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -output-using-type -emit-spirv-via-glsl
```
Comment (csyonghe) [inline tests/glsl-intrinsic/intrinsic-basic.slang:8]: If you want a test to run with both glsl and direct spirv path, just remove `-emit-spirv-*` options and the test will be run twice. See `.github\workflows\windows-selfhosted.yml`.

**Pull request merged**